### PR TITLE
Function and FunctionReference documentation

### PR DESCRIPTION
**Pull request by NickNaso**
First pass on function documentation. Please review only *function.md*. I'm actually writing the function reference documentation and I will push it as soon as possible.

### CODE REVIEW

File: doc/function.md [hunk 1]
```diff
@@ -1,5 +1,255 @@
 # Function
 
-You are reading a draft of the next documentation and it's in continuous update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
+The **Function** class is a representation of the JavaScript function. A function
+is a JavaScript procedure, a set of statements that performs a task or calculates
+a value. This class provides some methods to create and exectue a function.
```
Comment (mhdawson) [inline doc/function.md]: How about
+The **Function** class is a representation of the JavaScript function. A function
+is a set of statements that performs a task or calculates
+a value. This class provides some methods to create and execute a function.

File: doc/function.md [hunk 2]
```diff
@@ -1,5 +1,255 @@
...
+
+  - bool
+  - Any integer type
+  - Any floating point type
+  - const char* (encoded using UTF-8, null-terminated)
+  - const char16_t* (encoded using UTF-16-LE, null-terminated)
+  - std::string (encoded using UTF-8)
+  - std::u16string
+  - napi::Value
+  - napi_value
```
Comment (mhdawson) [inline doc/function.md]: I don't think this is right?  It likely needs to be a napi_value which is handle for a JavaScript function.

File: doc/function.md [hunk 3]
```diff
@@ -1,5 +1,255 @@
...
+
+Creates instance of a `Function` object.
+
+```cpp
+template <typename Callable>
+static Function New(napi_env env, Callable cb, const char* utf8name = nullptr, void* data = nullptr);
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the Function object.
+- `[in] cb`: Object that implement the operator **()** accepting a `const CallbackInfo&`
```
Comment (mhdawson) [inline doc/function.md]: Should this just say it needs to be an instance of Callable ?

File: doc/function.md [hunk 4]
```diff
@@ -1,5 +1,255 @@
...
+### New
+
+```cpp
+template <typename Callable>
+static Function New(napi_env env, Callable cb, const std::string& utf8name, void* data = nullptr);
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the Function object.
+- `[in] cb`: Object that implement the operator **()** accepting a `const CallbackInfo&`
+and returns either void or Value.
```
Comment (mhdawson) [inline doc/function.md]: Same comment as above.

File: doc/function.md [hunk 5]
```diff
@@ -1,5 +1,255 @@
...
+### Call
+
+Calls a Javascript function from a native add-on.
+
+```cpp
+Value Call(napi_value recv, const std::vector<napi_value>& args) const;
+```
+
+- `[in] recv`: The `this` object passed to the called function.
+- `[in] args`: List of JavaScript values as `napi_value` representing the
```
Comment (mhdawson) [inline doc/function.md]: Maybe 

Vector of JavaScript values as `napi_value` representing the

File: doc/function.md [hunk 6]
```diff
@@ -1,5 +1,255 @@
...
+
+```cpp
+Value MakeCallback(napi_value recv, const std::initializer_list<napi_value>& args) const;
+```
+
+- `[in] recv`: The `this` object passed to the called function.
+- `[in] args`: Initializer list of JavaScript values as `napi_value` representing
+the arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
```
Comment (mhdawson) [inline doc/function.md]: I think we need to have somewhere, either in the intro or in the description of the MakeCallback function where you should use Call versus MakeCallback.  You should use Call if there is already a JavaScript function on the stack (for example when running a native method called from JavaScript), and only use MakeCallback in special cases like running OnError, OnOk in AsyncWorker which don't have an existing JavaScript function on the stack.

File: doc/function.md [hunk 7]
```diff
@@ -1,5 +1,255 @@
...
+```
+
+- `[in] recv`: The `this` object passed to the called function.
+- `[in] argc`: The number of the arguments passed to the function.
+- `[in] args`: Array of JavaScript values as `napi_value` representing the
+arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
+
+### New
```
Comment (mhdawson) [inline doc/function.md]: I think all the new methods should be together.

File: doc/function_reference.md [hunk 8]
```diff
@@ -1,5 +1,133 @@
...
+# FunctionReference
 
-You are reading a draft of the next documentation and it's in continuous update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
+## Methods
+
+### Constructor
+
+Creates a new
```
Comment (mhdawson) [inline doc/function_reference.md]: partial line.

File: doc/function_reference.md [hunk 9]
```diff
@@ -1,5 +1,133 @@
...
+
+```cpp
+FunctionReference();
+```
+
+- `[in] Env`: The environment in which to construct the
```
Comment (mhdawson) [inline doc/function_reference.md]: partial line. I see that in a bunch of places throughout.

File: doc/function_reference.md [hunk 10]
```diff
@@ -1,5 +1,133 @@
...
+
+## Operator
+
+```cpp
+Napi::Value operator ()(const std::initializer_list<napi_value>& args) const;
+```
+
+- `[in] Env`: The environment in which to construct the
+
+Returns a new
```
Comment (mhdawson) [inline doc/function_reference.md]: Function reference overall looks like its missing the substitution.  For example "Returns a new" should probably be "Returns a new FunctionReference"

Comment (mhdawson) [inline doc/function_reference.md]: I think we also need some explanation of why you use a FunctionReference instead of a Function and when.

File: doc/function.md [hunk 11]
```diff
@@ -1,5 +1,255 @@
...
+
+Creates a new empty instance of `Function`.
+
+```cpp
+Function();
+```
+
+### Constructor
+
+Creates a new instance of `Function` object.
```
Comment (gabrielschulhof) [inline doc/function.md]: of *the* `Function` object.

File: doc/function.md [hunk 12]
```diff
@@ -1,5 +1,255 @@
...
+  - std::string (encoded using UTF-8)
+  - std::u16string
+  - napi::Value
+  - napi_value
+
+Returns a non-empty `Function` instance.
+
+### New
+
+Creates instance of a `Function` object.
```
Comment (gabrielschulhof) [inline doc/function.md]: Creates *an* instance

File: doc/function.md (hunk 3 above)
Comment (gabrielschulhof) [reply doc/function.md]: implement*s*

File: doc/function.md [hunk 13]
```diff
@@ -1,5 +1,255 @@
...
+
+```cpp
+template <typename Callable>
+static Function New(napi_env env, Callable cb, const char* utf8name = nullptr, void* data = nullptr);
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the Function object.
+- `[in] cb`: Object that implement the operator **()** accepting a `const CallbackInfo&`
+and returns either void or Value.
```
Comment (gabrielschulhof) [inline doc/function.md]: Perhaps change this to "void or a `Napi::Value`".

File: doc/function.md [hunk 14]
```diff
@@ -1,5 +1,255 @@
...
+- `[in] utf8name`: Null-terminated string to be used as the name of the function.
+- `[in] data`: User-provided data context. This will be passed back into the
+function when invoked later.
+
+Returns instance of a `Function` object.
```
Comment (gabrielschulhof) [inline doc/function.md]: *an* instance

File: doc/function.md (hunk 4 above)
Comment (gabrielschulhof) [reply doc/function.md]: Ditto regarding "void or Value".

File: doc/function.md [hunk 15]
```diff
@@ -1,5 +1,255 @@
...
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the Function object.
+- `[in] cb`: Object that implement the operator **()** accepting a `const CallbackInfo&`
+and returns either void or Value.
+- `[in] utf8name`: String to be used as the name of the function.
+- `[in] data`: User-provided data context. This will be passed back into the
+function when invoked later.
+
+Returns instance of a `Function` object.
```
Comment (gabrielschulhof) [inline doc/function.md]: *an* instance

File: doc/function.md [hunk 16]
```diff
@@ -1,5 +1,255 @@
...
+Calls a Javascript function from a native add-on.
+
+```cpp
+Value Call(const std::initializer_list<napi_value>& args) const;
+```
+
+- `[in] args`: Initializer list of JavaScript values as `napi_value` representing
+the arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
```
Comment (gabrielschulhof) [inline doc/function.md]: "Returns a `Napi::Value` representing the JavaScript value..." we shouldn't say "JavaScript object" because that has a special meaning.

File: doc/function.md [hunk 17]
```diff
@@ -1,5 +1,255 @@
...
+Calls a Javascript function from a native add-on.
+
+```cpp
+Value Call(const std::vector<napi_value>& args) const;
+```
+
+- `[in] args`: List of JavaScript values as `napi_value` representing the
+arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
```
Comment (gabrielschulhof) [inline doc/function.md]: Ditto.

File: doc/function.md [hunk 18]
```diff
@@ -1,5 +1,255 @@
...
+
+```cpp
+Value Call(size_t argc, const napi_value* args) const;
+```
+
+- `[in] argc`: The number of the arguments passed to the function.
+- `[in] args`: Array of JavaScript values as `napi_value` representing the
+arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
```
Comment (gabrielschulhof) [inline doc/function.md]: Ditto also for all instances to follow.

File: doc/function.md [hunk 19]
```diff
@@ -1,5 +1,255 @@
...
+Value Call(napi_value recv, size_t argc, const napi_value* args) const;
+```
+
+- `[in] recv`: The `this` object passed to the called function.
+- `[in] argc`: The number of the arguments passed to the function.
+- `[in] args`: Array of JavaScript values as `napi_value` representing the
+arguments of the function.
+
+Returns a Value representing the JavaScript object returned by the function.
+
```
Comment (gabrielschulhof) [inline doc/function.md]: I'm surprised we don't have overloads which accept an array of `Napi::Value` items 😕

**Comment by NickNaso** [comment]
@mhdawson I'm starting working now.

File: doc/function.md (hunk 7 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 5 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 3 above)
Comment (NickNaso) [reply doc/function.md]: is it good for you: *Object that implements Callable.*?

File: doc/function.md (hunk 4 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 14 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 15 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 16 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 17 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 18 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 19 above)
Comment (NickNaso) [reply doc/function.md]: So do I, what do you think if I open an issue about that?  Just to remember this and maybe after the documentation I can try to work on that.

File: doc/function.md (hunk 13 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 11 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 12 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 2 above)
Comment (NickNaso) [reply doc/function.md]: Done.

File: doc/function.md (hunk 19 above)
Comment (NickNaso) [reply doc/function.md]: Good points. Thanks for sharing this.

**Comment by NickNaso** [comment]
Just tried to address all the things issued after first review.

**Comment by NickNaso** [comment]
@mhdawson They are both good to me. Thanks for your help.

**Comment by mhdawson** [comment]
@NickNaso thanks, will leave until tomorrow to let @gabrielschulhof time to take a look and then will go ahead and land.

File: doc/function.md [hunk 20]
```diff
@@ -1,5 +1,282 @@
 # Function
 
-You are reading a draft of the next documentation and it's in continuous update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
+The **Function** class provides a set of methods to create a function object in
```
Comment (gabrielschulhof) [inline doc/function.md]: "... a set of methods ~~to create~~*for creating* ..."

File: doc/function.md [hunk 21]
```diff
@@ -1,5 +1,282 @@
...
+native code that can later be called from JavaScript. The created function is not
+automatically visible from JavaScript, instead it needs to be part of the add-on's
```
Comment (gabrielschulhof) [inline doc/function.md]: This is a run-on sentance. Let's have "... from JavaScript. Instead it ..."

File: doc/function.md [hunk 22]
```diff
@@ -1,5 +1,282 @@
...
+module exports or be returned by one of the modules exported functions.
```
Comment (gabrielschulhof) [inline doc/function.md]: "module's" instead of "modules"

File: doc/function.md [hunk 23]
```diff
@@ -1,5 +1,282 @@
...
+
+In addition the `Function` class also provides methods that can be used to call
+functions that were created in JavaScript and passed to the native.
```
Comment (gabrielschulhof) [inline doc/function.md]: "... to the native *add-on*", perhaps?

File: doc/function.md [hunk 24]
```diff
@@ -1,5 +1,282 @@
...
+```
+
+The above code can be used from JavaScript as follows:
+
+```js
+const addon = require('./addon');
+addon.fn();
+```
+
+The `Function` class allows to call a JavaScript function object from a native
```
Comment (gabrielschulhof) [inline doc/function.md]: Maybe we should rephrase this as "With the `Function` class it is possible to call..."

File: doc/function.md [hunk 25]
```diff
@@ -1,5 +1,282 @@
...
+add-on with two different methods: `Call` and `MackeCallback`.
```
Comment (gabrielschulhof) [inline doc/function.md]: `MakeCallback`

File: doc/function.md [hunk 26]
```diff
@@ -1,5 +1,282 @@
...
+The API of these two methods are very similar, but they are used in different context.
```
Comment (gabrielschulhof) [inline doc/function.md]: "*is* very similar, ..."

File: doc/function.md [hunk 27]
```diff
@@ -1,5 +1,282 @@
...
+`MakeCallback` method is used to call from native code back into JavaScript after
```
Comment (gabrielschulhof) [inline doc/function.md]: "*The* `MakeCallback` method..."

File: doc/function.md [hunk 28]
```diff
@@ -1,5 +1,282 @@
...
+returning from an [asynchronous operation](async_operations.md) and in general in
+situations which don't have an existing JavaScript function on the stack.
+`Call` method is used when there is already a JavaScript function on the stack
```
Comment (gabrielschulhof) [inline doc/function.md]: "*The* `Call` method..."

File: doc/function.md (hunk 16 above)
Comment (gabrielschulhof) [reply doc/function.md]: NP 🙂 Weird though ... github hasn't updated to show that 😕

Comment (gabrielschulhof) [reply doc/function.md]: Oh, I see - you fixed the ones below, but you missed this one. NP, but please say "&#96;Napi::Value&#96;" rather than merely "&#96;Value&#96;" – that is, please use the name space with the type name.

File: doc/function.md [hunk 29]
```diff
@@ -1,5 +1,282 @@
...
+```
+
+- `[in] args`: Vector of JavaScript values as `napi_value` representing the
+arguments of the function.
+
+Returns a `Value` representing the JavaScript value returned by the function.
+
+### Call
+
+Calls a Javascript function from a native add-on.
```
Comment (gabrielschulhof) [inline doc/function.md]: "JavaScript"

File: doc/function.md [hunk 30]
```diff
@@ -1,5 +1,282 @@
...
+
+```cpp
+Value Call(size_t argc, const napi_value* args) const;
```
Comment (gabrielschulhof) [inline doc/function.md]: Hmmm, come to think of it, should the return type not also be given as `Napi::Value`? I thought we had agreed to be explicit about the name space.

File: doc/function_reference.md [hunk 31]
```diff
@@ -1,5 +1,193 @@
-# Function reference
+# FunctionReference
 
-You are reading a draft of the next documentation and it's in continuous update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
+FunctionReference is a subclass of [Reference](reference.md), and is equivalent to
```
Comment (gabrielschulhof) [inline doc/function_reference.md]: The first word should be in backticks.

File: doc/function_reference.md [hunk 32]
```diff
@@ -1,5 +1,193 @@
...
+an instance of `Reference<Function>`. This means that a FunctionReference holds a
```
Comment (gabrielschulhof) [inline doc/function_reference.md]: "FunctionReference" should be in backticks.

File: doc/function_reference.md [hunk 33]
```diff
@@ -1,5 +1,193 @@
-# Function reference
+# FunctionReference
 
-You are reading a draft of the next documentation and it's in continuous update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
+FunctionReference is a subclass of [Reference](reference.md), and is equivalent to
+an instance of `Reference<Function>`. This means that a FunctionReference holds a 
+[Function](function.md), and a count of the number of references to that Function.
```
Comment (gabrielschulhof) [inline doc/function_reference.md]: Backticks around "Function" (both in the link and at the end of the line).

File: doc/function_reference.md [hunk 34]
```diff
@@ -1,5 +1,193 @@
-# Function reference
+# FunctionReference
 
-You are reading a draft of the next documentation and it's in continuous update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
+FunctionReference is a subclass of [Reference](reference.md), and is equivalent to
+an instance of `Reference<Function>`. This means that a FunctionReference holds a 
+[Function](function.md), and a count of the number of references to that Function. 
+When the count is greater than 0, a FunctionReference is not eligible for garbage
```
Comment (gabrielschulhof) [inline doc/function_reference.md]: Backticks around `FunctionReference` please!

Comment (gabrielschulhof) [reply doc/function_reference.md]: In general, please make another pass and add backticks around type names etc.!

File: doc/function_reference.md [hunk 35]
```diff
@@ -1,5 +1,193 @@
...
+
+- `[in] args`: Initializer list of JavaScript values as napi_value representing 
+the arguments of the contructor function.
+
+Returns a new JavaScript object.
+
+### New
+
+Creates a new JavaScript value from one that represents the constructor for the
+object.
```
Comment (gabrielschulhof) [inline doc/function_reference.md]: Perhaps "Constructs a new instance by calling the constructor held by this reference."?

File: doc/function_reference.md [hunk 36]
```diff
@@ -1,5 +1,193 @@
...
+
+```cpp
+Object New(const std::vector<napi_value>& args) const;
+```
+
+- `[in] args`: Vector of JavaScript values as napi_value representing the 
+arguments of the constructor function.
+
+Returns a new JavaScript object.
```
Comment (gabrielschulhof) [inline doc/function_reference.md]: Ditto ... ?

**Comment by NickNaso** [comment]
@gabrielschulhof Could you take a look at my last update? Thanks :-)

File: doc/function.md [hunk 37]
```diff
@@ -1,5 +1,282 @@
...
-You are reading a draft of the next documentation and it's in continuous update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
+The **Function** class provides a set of methods for creating a function object in
+native code that can later be called from JavaScript. The created function is not
+automatically visible from JavaScript. Instead it needs to be part of the add-on's
+module exports or be returned by one of the module's exported functions.
+
+In addition the `Function` class also provides methods that can be used to call
+functions that were created in JavaScript and passed to the native  add-on.
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function.md]: Done!

File: doc/function.md [hunk 38]
```diff
@@ -1,5 +1,282 @@
...
+
+```js
+const addon = require('./addon');
+addon.fn();
+```
+
+With the `Function` class it is possible to call a JavaScript function object
+from a native add-on with two different methods: `Call` and `MakeCallback`.
+The API of these two methods is very similar, but they are used in different
+context. The `MakeCallback` method is used to call from native code back into
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function.md]: Done!

File: doc/function.md [hunk 39]
```diff
@@ -1,5 +1,282 @@
...
+### Constructor
+
+Creates a new instance of the `Function` object.
+
+```cpp
+Function(napi_env env, napi_value value);
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the Value object.
+- `[in] value`: The `napi_value` which is handle for a JavaScript function.
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function.md]: Done!

File: doc/function.md [hunk 40]
```diff
@@ -1,5 +1,282 @@
...
+### New
+
+Creates an instance of a `Function` object.
+
+```cpp
+template <typename Callable>
+static Function New(napi_env env, Callable cb, const char* utf8name = nullptr, void* data = nullptr);
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the Function object.
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function.md]: Done!

File: doc/function_reference.md [hunk 41]
```diff
@@ -1,5 +1,191 @@
-# Function reference
+# FunctionReference
 
-You are reading a draft of the next documentation and it's in continuous update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
+`FunctionReference` is a subclass of [Reference](reference.md), and is equivalent to
+an instance of `Reference<Function>`. This means that a FunctionReference holds a
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function_reference.md]: Done!

File: doc/function_reference.md [hunk 42]
```diff
@@ -1,5 +1,191 @@
...
+[`Function`](function.md), and a count of the number of references to that `Function`.
+When the count is greater than 0, a FunctionReference is not eligible for garbage
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function_reference.md]: Done!

File: doc/function_reference.md [hunk 43]
```diff
@@ -1,5 +1,191 @@
...
+collection. This ensures that the `Function` will remain accessible, even if the
+original reference to it is no longer available.
+`FunctionReference` allows the referenced JavaScript function object  to be called
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function_reference.md]: Done!

File: doc/function_reference.md [hunk 44]
```diff
@@ -1,5 +1,191 @@
...
+
+The `FunctionReference` class inherits its behavior from the `Reference` class
+(for more info see: [Reference](reference.md)).
+
+## Methods
+
+### Weak
+
+Creates a "weak" reference to the value, in that the initial count of number of
+references is set to 0.
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function_reference.md]: Done!

File: doc/function_reference.md [hunk 45]
```diff
@@ -1,5 +1,191 @@
...
+
+### New
+
+Constructs a new instance by calling the constructor held by this reference.
+
+```cpp
+Napi::Object New(const std::initializer_list<napi_value>& args) const;
+```
+
+- `[in] args`: Initializer list of JavaScript values as napi_value representing
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function_reference.md]: Done!

File: doc/function_reference.md [hunk 46]
```diff
@@ -1,5 +1,191 @@
...
+
+### New
+
+Constructs a new instance by calling the constructor held by this reference.
+
+```cpp
+Napi::Object New(const std::vector<napi_value>& args) const;
+```
+
+- `[in] args`: Vector of JavaScript values as napi_value representing the
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function_reference.md]: Done!

**Comment by mhdawson** [comment]
@gabrielschulhof I think this is waiting for you now.

File: doc/function_reference.md [hunk 47]
```diff
@@ -1,5 +1,191 @@
...
+```
+
+- `[in] value`: The value which is to be referenced.
+
+Returns the newly created reference.
+
+### Persistent
+
+Creates a "persistent" reference to the value, in that the initial count of
+number of references is set to 1.
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function_reference.md]: Done!

**Comment by gabrielschulhof** [comment]
I actually have to take a few hours before I can review. BRB 🙂

File: doc/function.md [hunk 48]
```diff
@@ -1,5 +1,282 @@
...
+Returns an instance of a `Function` object.
+
+### New
+
+```cpp
+template <typename Callable>
+static Function New(napi_env env, Callable cb, const std::string& utf8name, void* data = nullptr);
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the Function object.
```
Comment (gabrielschulhof) [inline doc/function.md]: Backticks around "Function"

File: doc/function.md [hunk 49]
```diff
@@ -1,5 +1,282 @@
...
+Creates a new instance of the `Function` object.
+
+```cpp
+Function(napi_env env, napi_value value);
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the Value object.
+- `[in] value`: The `napi_value` which is a handle for a JavaScript function.
+
+Returns a non-empty `Function` instance.
```
Comment (gabrielschulhof) [inline doc/function.md]: We should really be consistent about whether we say `Napi::Value`, `Napi::Function`, `Napi::FunctionReference`, etc. or `Value`, `Function`, or `FunctionReference`. IMO we should always add the name space in front of the class name. @mhdawson what do you think?

File: doc/function_reference.md [hunk 50]
```diff
@@ -1,5 +1,191 @@
-# Function reference
+# FunctionReference
 
-You are reading a draft of the next documentation and it's in continuous update so
-if you don't find what you need please refer to:
-[C++ wrapper classes for the ABI-stable C APIs for Node.js](https://nodejs.github.io/node-addon-api/)
+`FunctionReference` is a subclass of [Reference](reference.md), and is equivalent to
```
Comment (gabrielschulhof) [inline doc/function_reference.md]: Here too, and in all cases where we use class names in the documentation, we should add the name space explicitly – that is, `Napi::FunctionReference` and `Napi::Reference` in this case. Additionally, "Reference" should be in backquotes.

File: doc/function_reference.md [hunk 51]
```diff
@@ -1,5 +1,191 @@
...
+When the count is greater than 0, a `FunctionReference` is not eligible for garbage
+collection. This ensures that the `Function` will remain accessible, even if the
+original reference to it is no longer available.
+`FunctionReference` allows the referenced JavaScript function object to be called
+from a native add-on with two different methods: `Call` and `MakeCallback`. See
+the documentation for [`Function`](function.md) for when `Call` should be used
+instead of `MakeCallback` and vice-versa.
+
+The `FunctionReference` class inherits its behavior from the `Reference` class
+(for more info see: [Reference](reference.md)).
```
Comment (gabrielschulhof) [inline doc/function_reference.md]: Backticks around "Reference"

File: doc/function_reference.md [hunk 52]
```diff
@@ -1,5 +1,191 @@
...
+### Constructor
+
+Creates a new instance of the `FunctionReference`.
+
+```cpp
+FunctionReference(napi_env env, napi_ref ref);
+```
+
+- `[in] env`: The environment in which to construct the FunctionReference object.
+- `[in] ref`: The N-API reference to be held by the FunctionReference.
```
Comment (gabrielschulhof) [inline doc/function_reference.md]: Backticks around "FunctionReference".

File: doc/function.md [hunk 53]
```diff
@@ -1,5 +1,282 @@
...
+
+### Constructor
+
+Creates a new instance of the `Function` object.
+
+```cpp
+Function(napi_env env, napi_value value);
+```
+
+- `[in] env`: The `napi_env` environment in which to construct the Value object.
```
[the remark this answers is not in the document]
Comment (NickNaso) [reply doc/function.md]: Done!

File: doc/function.md (hunk 48 above)
Comment (NickNaso) [reply doc/function.md]: Done!

File: doc/function_reference.md (hunk 52 above)
Comment (NickNaso) [reply doc/function_reference.md]: Done!

File: doc/function_reference.md (hunk 51 above)
Comment (NickNaso) [reply doc/function_reference.md]: Done!

File: doc/function_reference.md (hunk 50 above)
Comment (gabrielschulhof) [reply doc/function_reference.md]: @NickNaso the word "Reference" inside the square bracket also needs backticks.

**Comment by gabrielschulhof** [comment]
@mhdawson shall we go with full-name-space notation like `Napi::Value` throughout, or shall `Value` suffice? IMO we should go with full-name-space notation. Either way, we should be consistent throughout the docs.

**Comment by mhdawson** [comment]
I agree that full-name-space would be best.

**Comment by mhdawson** [comment]
ok, lets land and then update to always use the name-space

**Comment by mhdawson** [comment]
Landed as 757eb1f5a368d7f52af6faf279b3075d37cf337f
